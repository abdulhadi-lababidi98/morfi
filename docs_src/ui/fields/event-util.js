/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

type Event = { preventDefault: Function, target: { value: any } };
type EventHandler = Event => void;

type CustomInputHandling = { preventDefault?: boolean, falsyToUndefined?: boolean };

const inputHandler = (
    cb: any => mixed | void,
    { preventDefault = true, falsyToUndefined = false }: CustomInputHandling = {}
): EventHandler => {
    return event => {
        if (preventDefault) event.preventDefault();
        const original = event.target.value;
        cb(falsyToUndefined ? original || undefined : original);
    };
};

export const EventUtil = { inputHandler };
