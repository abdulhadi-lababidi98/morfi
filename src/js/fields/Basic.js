/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import type { ErrorMessage } from '../form';
import messages from '../../messages';

export const __ = ({ id, values }: ErrorMessage): string => {
    let result = messages[id];
    if (result && values) {
        result = Object.keys(values).reduce(
            (red: string, key: string) => red.replace(`{${key}}`, String(values[key])),
            result
        );
    }
    return result;
};

export const Label = ({ label, required = false }: { label: string, required?: boolean }) => {
    return <label className="control-label">{label + (required ? ' *' : '')}</label>;
};

export const DisplayError = ({ error }: { error: ErrorMessage }) => {
    return <span className="invalid-feedback">{__(error)}</span>;
};

type EventHandler = (SyntheticInputEvent<*>) => void;

export const onActionWrap = (cb?: any => void, preventDefault: boolean = true): EventHandler => {
    return event => {
        if (preventDefault) {
            event.preventDefault();
        }
        if (cb) {
            cb(event.target.value);
        }
    };
};
