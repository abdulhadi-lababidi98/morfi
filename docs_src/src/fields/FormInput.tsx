import React, { Component } from 'react';
import { DisplayError, Label, onActionWrap } from './Basic';
import { Eye } from '../icons/Eye';
import { Spinner } from '../icons/Spinner';
import { Morfi, FieldControls, FormField } from '../../../src';

type AllowedTypes = 'text' | 'password';

type CommonInputProps = {
    label?: string;
    className?: string;
    autoFocus?: boolean;
    disabled?: boolean;
    type?: AllowedTypes;
    placeholder?: string;
    pending?: boolean;
};

type InputProps = FieldControls<string> & CommonInputProps;

export class Input extends Component<InputProps, { showPassword: boolean }> {
    state = { showPassword: false };
    showPasswordToggle = () => this.setState((state) => ({ showPassword: !state.showPassword }));
    type = () => (this.props.type !== 'password' ? this.props.type : this.state.showPassword ? 'text' : 'password');
    render() {
        const {
            value = '',
            label,
            error,
            required,
            className = 'form-group',
            onChange,
            onBlur,
            autoFocus,
            disabled,
            type = 'text',
            placeholder,
            pending,
        } = this.props;
        return (
            <div className={className}>
                {label && <Label {...{ label, required }} />}
                <input
                    onChange={onActionWrap(onChange)}
                    onBlur={onActionWrap(onBlur)}
                    disabled={disabled}
                    className={'form-control' + (error ? ' is-invalid' : '')}
                    type={this.type()}
                    {...{ placeholder, value, autoFocus }}
                />
                {type === 'password' && (
                    <span onClick={this.showPasswordToggle}>
                        <Eye stroked={this.state.showPassword} />
                    </span>
                )}
                {pending && <Spinner />}
                {error && <DisplayError error={error} />}
            </div>
        );
    }
}

type FormInputProps = { field: FormField<string> } & CommonInputProps;

export const FormInput: React.FC<FormInputProps> = ({ field, ...rest }) => {
    const fieldProps = Morfi.useField(field);
    return <Input {...fieldProps} {...rest} />;
};
