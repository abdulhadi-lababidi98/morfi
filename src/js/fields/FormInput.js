/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React, { Component } from 'react';
import { Error, Label, onActionWrap } from './Basic';
import { Field } from '../form';
import type { _ErrorMessage } from '../form/Form-classes';
import { Spinner } from '../icons/Spinner';
import { Eye } from '../icons/Eye';

type AllowedTypes = 'text' | 'password';

type InputProps = {
    value?: string,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: string => void,
    onBlur?: string => void,
    autoFocus?: boolean,
    disabled?: boolean,
    type?: AllowedTypes,
    placeholder?: string,
    pending?: boolean,
};

export class Input extends Component<InputProps, { showPassword: boolean }> {
    state = { showPassword: false };
    showPasswordToggle = () => this.setState(state => ({ showPassword: !state.showPassword }));
    type = () => (this.props.type !== 'password' ? this.props.type : this.state.showPassword ? 'text' : 'password');
    render() {
        const {
            value = '',
            label,
            error,
            required,
            className,
            onChange,
            onBlur,
            autoFocus,
            disabled,
            type = 'text',
            placeholder,
            pending,
        } = this.props;
        return (
            <div className={className}>
                {label && <Label {...{ label, required }} />}
                <input
                    onChange={onActionWrap(onChange)}
                    onBlur={onActionWrap(onBlur)}
                    disabled={disabled || pending}
                    className={'form-control' + (error ? ' is-invalid' : '')}
                    type={this.type()}
                    {...{ placeholder, value, autoFocus }}
                />
                {type === 'password' && (
                    <span onClick={this.showPasswordToggle}>
                        <Eye stroked={this.state.showPassword} />
                    </span>
                )}
                {pending && <Spinner />}
                {error && <Error error={error} />}
            </div>
        );
    }
}

type FormInputProps = {|
    name: string,
    value?: string,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    disabled?: boolean,
    autoFocus?: boolean,
    type?: AllowedTypes,
    placeholder?: string,
    pending?: boolean,
|};

export const _FormInput = ({
    name,
    value,
    error,
    label,
    className = 'form-group',
    disabled,
    autoFocus,
    type,
    placeholder,
    pending,
}: FormInputProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <Input
                {...{
                    error,
                    value,
                    required,
                    onBlur,
                    onChange,
                    label,
                    className,
                    disabled,
                    autoFocus,
                    type,
                    placeholder,
                    pending,
                }}
            />
        )}
    </Field>
);
