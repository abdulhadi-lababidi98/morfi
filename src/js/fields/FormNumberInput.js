/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import { Error, Label, onActionWrap } from './Basic';
import { Field } from '../form';
import type { _ErrorMessage } from '../form/Form-classes';

type NumberInputProps = {
    value?: number,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: number => void,
    onBlur?: number => void,
    autoFocus?: boolean,
    disabled?: boolean,
};

export const NumberInput = ({
    value = 0,
    label,
    error,
    required,
    className,
    onChange,
    onBlur,
    autoFocus,
    disabled,
}: NumberInputProps) => {
    return (
        <div className={(className || '') + (error ? ' has-error' : '')}>
            {label && <Label {...{ label, required }} />}
            <input
                type="number"
                onChange={onActionWrap(onChange)}
                onBlur={onActionWrap(onBlur)}
                className="form-control"
                {...{ value, autoFocus, disabled }}
            />
            {error && <Error error={error} />}
        </div>
    );
};

type FormNumberInputProps = {|
    name: string,
    value?: number,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    disabled?: boolean,
    autoFocus?: boolean,
|};

export const _FormNumberInput = ({
    name,
    value,
    error,
    label,
    className = 'form-group',
    disabled,
    autoFocus,
}: FormNumberInputProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <NumberInput {...{ error, value, required, onBlur, onChange, label, className, disabled, autoFocus }} />
        )}
    </Field>
);
