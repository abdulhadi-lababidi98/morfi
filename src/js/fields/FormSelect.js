/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import type { _ErrorMessage } from '../form/Form-classes';
import { _Field as Field } from '../form/Field';
import { Error, Label, onActionWrap } from './Basic';

export type SelectOption = {| label: string, value: string |};

type SelectProps = {
    value?: string,
    label?: string,
    error?: _ErrorMessage,
    required?: boolean,
    className?: string,
    onChange?: string => void,
    onBlur?: string => void,
    options: Array<SelectOption>,
    disabled?: boolean,
};

const Option = ({ option }: { option: SelectOption }) => <option value={option.value}>{option.label}</option>;

const Select = ({ value, label, error, required, className, onChange, onBlur, options, disabled }: SelectProps) => {
    return (
        <div className={className}>
            {label && <Label {...{ label, required }} />}
            <select
                className="form-control"
                disabled={disabled}
                onChange={onActionWrap(onChange)}
                onBlur={onActionWrap(onBlur)}
                value={value}>
                {options.map((option: SelectOption, index: number) => <Option option={option} key={index} />)}
            </select>
            {error && <Error error={error} />}
        </div>
    );
};

type FormSelectProps = {|
    name: string,
    value?: string,
    error?: _ErrorMessage,
    label?: string,
    className?: string,
    options: Array<SelectOption>,
    disabled?: boolean,
|};

export const _FormSelect = ({
    name,
    value,
    error,
    options,
    label,
    className = 'form-group',
    disabled,
}: FormSelectProps) => (
    <Field name={name}>
        {({ onChange, onBlur, required }) => (
            <Select {...{ error, value, required, onBlur, onChange, label, className, disabled, options }} />
        )}
    </Field>
);
