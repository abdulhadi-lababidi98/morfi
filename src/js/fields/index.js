/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import { _FormInput as FormInput } from './FormInput';

export { FormInput };
