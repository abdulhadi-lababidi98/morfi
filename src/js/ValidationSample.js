/**
 * This file is part of form4react which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React, { Component } from 'react';
import { Validators } from './validators/validators';
import { Form } from './form';
import { FormInput } from './fields';
import type { FormData, FormValidation } from './form';
import DisplayValues from './DisplayValues';
import type { ValidationType } from './form/Form-classes';
import { Select } from './fields/FormSelect';
import type { SelectOption } from './fields/FormSelect';

const ValidationTypeOptions: Array<SelectOption> = [
    { label: 'onChange', value: 'onChange' },
    { label: 'onBlur', value: 'onBlur' },
    { label: 'onSubmit', value: 'onSubmit' },
];

const descriptionFor = (type: ValidationType) => {
    switch (type) {
        case 'onChange':
            return (
                'This validation type will automatically make a ' +
                'validation for each character you enter in the form fields.'
            );
        case 'onBlur':
            return 'This validation type will automatically make a validation each time you leave the form field.';
        case 'onSubmit':
            return (
                'This validation type will trigger the validators ' +
                'as soon as the form data was requested to be submitted.'
            );
        default:
            return '';
    }
};

type FormValues = {
    email?: string,
    pw?: string,
};

const validationFor = (type: ValidationType): FormValidation<FormValues> => {
    return {
        email: { [type]: Validators.email },
        pw: { [type]: Validators.regex({ re: /^[a-zA-Z0-9]{8,}$/, message: { id: 'validation.pw.requirements' } }) },
    };
};

type ValidationSampleState = {
    data: FormData<FormValues>,
    validation: FormValidation<FormValues>,
    currentValidationType: ValidationType,
};

const initialState = {
    data: { values: {}, errors: {} },
    validation: validationFor('onChange'),
    currentValidationType: 'onChange',
};

export class ValidationSample extends Component<{}, ValidationSampleState> {
    state: ValidationSampleState = initialState;

    onChange = (data: FormData<FormValues>) => this.setState({ data });

    onSubmit = (): Promise<void> => {
        // simulate server request
        return window.sleep(2000);
    };

    chooseValidationType = (type: any /* ValidationType <- is a problem with the used select component */) => {
        this.setState({ validation: validationFor(type), currentValidationType: type });
    };

    render(): React$Node {
        const data = this.state.data;
        const { values, errors, submitting } = data;
        return (
            <div className="col-12">
                <Select
                    label="Choose the validation type"
                    className="validation-type-picker mt-3 mb-2 col-md-6 row"
                    options={ValidationTypeOptions}
                    value={this.state.currentValidationType}
                    onChange={this.chooseValidationType}
                />
                <p className="small font-italic">{descriptionFor(this.state.currentValidationType)}</p>
                <hr />
                <Form validation={this.state.validation} onChange={this.onChange} data={data} onSubmit={this.onSubmit}>
                    <div className="row">
                        <div className="col-md-6">
                            <FormInput
                                name="email"
                                value={values.email}
                                error={errors.email}
                                label="Email"
                                placeholder="Please enter your email address"
                            />
                            <FormInput
                                name="pw"
                                type="password"
                                value={values.pw}
                                error={errors.pw}
                                label="Password"
                                placeholder="Set your password"
                            />
                            <div className="btn-toolbar">
                                <button className="btn btn-success" disabled={submitting || Form.hasErrors(data)}>
                                    {submitting && <i className="fa fa-circle-notch fa-spin" />} Submit
                                </button>
                            </div>
                        </div>
                        <div className="col-md-6 mt-3">
                            <DisplayValues data={this.state.data} />
                        </div>
                    </div>
                </Form>
            </div>
        );
    }
}
