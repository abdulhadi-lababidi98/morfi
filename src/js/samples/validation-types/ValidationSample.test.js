/**
 * This file is part of morfi which is released under MIT license.
 *
 * The LICENSE file can be found in the root directory of this project.
 *
 * @flow
 */

import React from 'react';
import { ValidationSample } from './ValidationSample';
import { shallow } from 'enzyme';
import { mountForm } from '../../../test/form-test-util';
import { Select } from '../../fields/FormSelect';

describe('AsyncValidationSample', () => {
    it('renders the initial form', () => {
        expect(shallow(<ValidationSample />)).toMatchSnapshot();
    });

    it('validates the form integration with onChange - Validations', () => {
        mountForm(<ValidationSample />)
            .change('email', 'nomail.com')
            .assume({ values: { email: 'nomail.com' }, errors: { email: { id: 'validation.email.requirements' } } })
            .change('email', 'some@mail.com')
            .assume({ values: { email: 'some@mail.com' }, errors: { email: undefined } })
            .change('pw', '1234567')
            .assumeErrors({ pw: { id: 'validation.pw.requirements' } })
            .change('pw', '12345678')
            .assumeErrors({ pw: undefined });
    });

    it('validates the form integration with onBlur - Validations', () => {
        const wrapper = mountForm(<ValidationSample />);
        wrapper
            .mounted()
            .find(Select)
            .props()
            .onChange('onBlur');
        wrapper
            .change('email', 'nomail.com')
            .assume({ values: { email: 'nomail.com' }, errors: { email: undefined } })
            .blur()
            .assume({ values: { email: 'nomail.com' }, errors: { email: { id: 'validation.email.requirements' } } })
            .change('email', 'some@mail.com')
            .assume({ values: { email: 'some@mail.com' }, errors: { email: undefined } })
            .blur()
            .assume({ values: { email: 'some@mail.com' }, errors: { email: undefined } })
            .change('pw', '1234567')
            .assume({ errors: { pw: undefined } })
            .blur()
            .assume({ errors: { pw: { id: 'validation.pw.requirements' } } })
            .change('pw', '12345678')
            .assume({ errors: { pw: undefined } })
            .blur()
            .assume({ errors: { pw: undefined } });
    });

    it('validates the form integration with onSubmit - Validations', async () => {
        const wrapper = mountForm(<ValidationSample />);
        wrapper
            .mounted()
            .find(Select)
            .props()
            .onChange('onSubmit');
        wrapper
            .change('email', 'nomail.com')
            .change('pw', '1234567')
            .blur()
            .assumeErrors({ email: undefined, pw: undefined });

        await wrapper.submit().nextTick();

        wrapper
            .assumeErrors({
                email: { id: 'validation.email.requirements' },
                pw: { id: 'validation.pw.requirements' },
            })
            .change('email', 'some@mail.com')
            .change('pw', '12345678')
            .assumeErrors({ email: undefined, pw: undefined });

        await wrapper.submit().nextTick();

        wrapper.assumeErrors({ email: undefined, pw: undefined });
    });
});
