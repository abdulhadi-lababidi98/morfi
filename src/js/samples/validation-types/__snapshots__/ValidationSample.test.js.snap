// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncValidationSample renders the initial form 1`] = `
<div
  className="col-12"
>
  <Component
    className="validation-type-picker mt-3 mb-2 col-md-6 row"
    label="Choose the validation type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "onChange",
          "value": "onChange",
        },
        Object {
          "label": "onBlur",
          "value": "onBlur",
        },
        Object {
          "label": "onSubmit",
          "value": "onSubmit",
        },
      ]
    }
    value="onChange"
  />
  <p
    className="small font-italic"
  >
    This validation type will automatically make a validation for each character you enter in the form fields.
  </p>
  <hr />
  <_Form
    data={
      Object {
        "errors": Object {},
        "values": Object {},
      }
    }
    onChange={[Function]}
    onSubmit={[Function]}
    validation={
      Object {
        "email": Object {
          "onChange": [Function],
        },
        "pw": Object {
          "onChange": [Function],
        },
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <Component
          label="Email"
          name="email"
          placeholder="Please enter your email address"
        />
        <Component
          label="Password"
          name="pw"
          placeholder="Set your password"
          type="password"
        />
        <div
          className="btn-toolbar"
        >
          <button
            className="btn btn-success"
            disabled={false}
          >
            <span>
               Submit
            </span>
          </button>
        </div>
      </div>
      <div
        className="col-md-6 mt-3"
      >
        <DisplayValues
          data={
            Object {
              "errors": Object {},
              "values": Object {},
            }
          }
        />
      </div>
    </div>
  </_Form>
</div>
`;
